### ISSUE TITLE
current_version fields vs project.version field in pyproject.toml

### ISSUE THREAD

**Issue by ddewaele** [callowayproject/bump-my-version#215]
We are currently using the bump-my-version config as outlined in the [README](https://github.com/callowayproject/bump-my-version?tab=readme-ov-file#add-support-for-pre-release-versions).

Our `pyproject.toml` looks like this 

```
[project]
name = "acmecorp-libraryA"
version = "0.0.1"

[tool.bumpversion]
current_version = "0.0.1"
parse = """(?x)
    (?P<major>0|[1-9]\\d*)\\.
    (?P<minor>0|[1-9]\\d*)\\.
    (?P<patch>0|[1-9]\\d*)
    (?:
        -                             # dash separator for pre-release section
        (?P<pre_l>[a-zA-Z-]+)         # pre-release label
        (?P<pre_n>0|[1-9]\\d*)        # pre-release version number
    )?                                # pre-release section is optional
"""
serialize = [
    "{major}.{minor}.{patch}-{pre_l}{pre_n}",
    "{major}.{minor}.{patch}",
]
search = "{current_version}"
replace = "{new_version}"
regex = true
ignore_missing_version = false
ignore_missing_files = false
tag = true
sign_tags = false
tag_name = "v{new_version}"
tag_message = "[skip ci] Bump version: {current_version} → {new_version}"
allow_dirty = false
commit = true
message = "[skip ci] Bump version: {current_version} → {new_version}"
commit_args = ""
```

All of our libraries will folllow the same versioning scheme and we would like to externalize the bump-my-version config in a separate `bumpversion.toml` file to keep our `pyproject.toml` files clean and re-use the bump-my-version config for all projects.

We are currently doing our version bumps using 

```
bump-my-version bump patch pyproject.toml
```

This does all of our tagging, updates the pyproject.toml file `project.version` as well as the `tool.bumpversion.current_version`

However, i am a bit puzzled by the apparent redundant `current_version` field in the `[tool.bumpversion]` or in the `bumpversion.toml`.

Is there a reason why a separate (duplicate) field this is used instead of just using relying on `project.version` as specified in the `pyproject.toml` ? 

Is there a way to have bump-my-version just use the `project.version` as a reference to keep track of the current version and update that field when it bumps a version ? (I don't see a use-case where these 2 fields would ever get out-of-sync).

Without the `current_version` in the config file the bump-my-version will fail with this

```
toolbox-iterator ❯ bump-my-version bump patch pyproject.toml
Traceback (most recent call last):
  File "/opt/miniconda3/envs/toolbox-iterator/bin/bump-my-version", line 8, in <module>
    sys.exit(cli())
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/rich_click/rich_command.py", line 367, in __call__
    return super().__call__(*args, **kwargs)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/rich_click/rich_command.py", line 152, in main
    rv = self.invoke(ctx)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/bumpversion/cli.py", line 309, in bump
    do_bump(version_part, new_version, config, found_config_file, dry_run)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/bumpversion/bump.py", line 108, in do_bump
    update_config_file(config_file, config, version, next_version, ctx, dry_run)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/bumpversion/config/files.py", line 150, in update_config_file
    updater.update_file(current_version, new_version, context, dry_run)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/bumpversion/files.py", line 340, in update_file
    self._update_toml_file(search_for, raw_search_pattern, replace_with, dry_run)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/bumpversion/files.py", line 349, in _update_toml_file
    value_before = get_nested_value(toml_data, self.file_change.key_path)
  File "/opt/miniconda3/envs/toolbox-iterator/lib/python3.9/site-packages/bumpversion/utils.py", line 70, in get_nested_value
    raise KeyError(f"Key '{key}' not found at '{'.'.join(keys[:keys.index(key)])}'")
KeyError: "Key 'current_version' not found at 'tool.bumpversion'"
```

Because the `current_version` is obviously project specific, this currently prevents us from have 1 global `bumpversion.toml` that all our projects could use.

**Comment by coordt** [comment]
The primary reason is that the `project.version` does not need to be a valid string. The [`dynamic`](https://packaging.python.org/en/latest/specifications/pyproject-toml/#dynamic) key allows for you to specify it dynamically. [Here is an example for Hatch.](https://hatch.pypa.io/latest/version/)

However, if the `current_version` is missing, it will attempt to get the version from the latest tag, and will not add back a `current_version` key.

**Comment by ddewaele** [comment]
My apologies. 

I was still on `0.24.2` where I was getting the above stack trace if `current_version` was not present in the config. 

Seems like it is fixed in `0.24.3` 

Great job on the library !
